File: ports/go/capabilities.go
package hqtui

import (
	"os"
	"strings"
)

// What the terminal can actually do. Detection is deliberately conservative: we
// degrade colors and glyphs rather than print mojibake on someone's console.

type ColorDepth int

const (
	ColorTrueColor ColorDepth = iota
	ColorAnsi256
	ColorAnsi16
	ColorNone
)

func (c ColorDepth) String() string {
	switch c {
	case ColorTrueColor:
		return "truecolor"
	case ColorAnsi256:
		return "ansi256"
	case ColorAnsi16:
		return "ansi16"
	}
	return "none"
}

func ParseColorDepth(s string) (ColorDepth, bool) {
	switch s {
	case "truecolor":
		return ColorTrueColor, true
	case "ansi256":
		return ColorAnsi256, true
	case "ansi16":
		return ColorAnsi16, true
	case "none":
		return ColorNone, true
	}
	return ColorTrueColor, false
}

type Capabilities struct {
	// TTY reports whether stdout is a real terminal, not a pipe or a file.
	TTY       bool
	Colors    ColorDepth
	TrueColor bool
	Unicode   bool
	Braille   bool
	Mouse     bool
	// SynchronizedOutput is DEC 2026 atomic frame updates.
	SynchronizedOutput bool
	BracketedPaste     bool
	FocusEvents        bool
	Tmux               bool
	Screen             bool
	SSH                bool
	Windows            bool
	// Program is a best guess at the emulator: kitty, wezterm, ghostty, iterm,
	// alacritty, vscode, windows-terminal, xterm, unknown.
	Program string
}

// CapabilityOverrides forces a capability rather than detecting it. Used by
// tests and by apps that know better than the environment does.
type CapabilityOverrides struct {
	Colors             *ColorDepth
	Unicode            *bool
	Braille            *bool
	Mouse              *bool
	SynchronizedOutput *bool
	TTY                *bool
}

// Env is the environment as a map, so detection is testable without touching
// the real process environment.
type Env map[string]string

func ProcessEnv() Env {
	env := Env{}
	for _, kv := range os.Environ() {
		if k, v, ok := strings.Cut(kv, "="); ok {
			env[k] = v
		}
	}
	return env
}

func (e Env) get(key string) string  { return e[key] }
func (e Env) has(key string) bool    { return e[key] != "" }

func DetectProgram(env Env) string {
	term := env.get("TERM")
	program := env.get("TERM_PROGRAM")
	switch {
	case env.has("KITTY_WINDOW_ID") || term == "xterm-kitty":
		return "kitty"
	case env.has("WEZTERM_EXECUTABLE") || program == "WezTerm":
		return "wezterm"
	case env.has("GHOSTTY_RESOURCES_DIR") || term == "xterm-ghostty":
		return "ghostty"
	case program == "iTerm.app":
		return "iterm"
	case env.has("ALACRITTY_WINDOW_ID") || term == "alacritty":
		return "alacritty"
	case program == "vscode":
		return "vscode"
	case env.has("WT_SESSION"):
		return "windows-terminal"
	case program == "Apple_Terminal":
		return "apple-terminal"
	case env.has("KONSOLE_VERSION"):
		return "konsole"
	case strings.HasPrefix(term, "xterm"):
		return "xterm"
	}
	return "unknown"
}

var truecolorPrograms = map[string]bool{
	"kitty": true, "wezterm": true, "ghostty": true, "iterm": true,
	"vscode": true, "windows-terminal": true, "konsole": true,
}

func detectColors(env Env, tty bool) ColorDepth {
	if env.has("NO_COLOR") {
		return ColorNone
	}
	switch env.get("FORCE_COLOR") {
	case "0", "false":
		return ColorNone
	case "1":
		return ColorAnsi16
	case "2":
		return ColorAnsi256
	case "3":
		return ColorTrueColor
	}
	// Any other non-empty value — FORCE_COLOR=true is the common one — asserts
	// that color works. It is a floor, not a ceiling: returning a level here
	// would cap a truecolor terminal at 16 colors. It only waives the tty check.
	forced := env.has("FORCE_COLOR")
	if !tty && !forced {
		return ColorNone
	}
	term := env.get("TERM")
	if term == "dumb" {
		return ColorNone
	}
	switch strings.ToLower(env.get("COLORTERM")) {
	case "truecolor", "24bit":
		return ColorTrueColor
	}
	if truecolorPrograms[DetectProgram(env)] {
		return ColorTrueColor
	}
	if strings.Contains(term, "256") {
		return ColorAnsi256
	}
	return ColorAnsi16
}

func detectUnicode(env Env, windows bool) bool {
	// A dumb terminal has no glyph repertoire to speak of. The Linux console is
	// not in that category — its default font draws box and block elements
	// perfectly well — so only Braille is withheld from it, below.
	if env.get("TERM") == "dumb" {
		return false
	}
	locale := env.get("LC_ALL")
	if locale == "" {
		locale = env.get("LC_CTYPE")
	}
	if locale == "" {
		locale = env.get("LANG")
	}
	upper := strings.ToUpper(locale)
	if strings.Contains(upper, "UTF8") || strings.Contains(upper, "UTF-8") {
		return true
	}
	// Windows Terminal and modern emulators are UTF-8 regardless of locale vars.
	if env.has("WT_SESSION") || env.has("TERM_PROGRAM") || env.has("KITTY_WINDOW_ID") {
		return true
	}
	if windows {
		return env.has("WT_SESSION")
	}
	return locale == ""
}

var syncPrograms = map[string]bool{
	"kitty": true, "wezterm": true, "ghostty": true, "iterm": true,
	"windows-terminal": true, "konsole": true, "alacritty": true,
}

func DetectCapabilities(o CapabilityOverrides) Capabilities {
	return DetectCapabilitiesIn(o, ProcessEnv(), isTerminal())
}

func DetectCapabilitiesIn(o CapabilityOverrides, env Env, isTTY bool) Capabilities {
	tty := isTTY
	if o.TTY != nil {
		tty = *o.TTY
	}
	term := env.get("TERM")
	program := DetectProgram(env)
	tmux := env.has("TMUX") || strings.HasPrefix(term, "tmux") || strings.HasPrefix(term, "screen")
	screen := strings.HasPrefix(term, "screen") && !env.has("TMUX")
	ssh := env.has("SSH_CLIENT") || env.has("SSH_TTY") || env.has("SSH_CONNECTION")

	colors := detectColors(env, tty)
	if o.Colors != nil {
		colors = *o.Colors
	}
	windows := runningOnWindows()
	unicode := detectUnicode(env, windows)
	if o.Unicode != nil {
		unicode = *o.Unicode
	}
	// The Linux console draws box and block elements but has no Braille in its
	// default font, which is the one glyph class it genuinely lacks.
	braille := unicode && program != "apple-terminal" && term != "linux"
	if o.Braille != nil {
		braille = *o.Braille
	}
	mouse := tty && term != "dumb" && term != "linux"
	if o.Mouse != nil {
		mouse = *o.Mouse
	}
	sync := tty && (syncPrograms[program] || tmux)
	if o.SynchronizedOutput != nil {
		sync = *o.SynchronizedOutput
	}

	return Capabilities{
		TTY:                tty,
		Colors:             colors,
		TrueColor:          colors == ColorTrueColor,
		Unicode:            unicode,
		Braille:            braille,
		Mouse:              mouse,
		SynchronizedOutput: sync,
		BracketedPaste:     tty && term != "dumb",
		FocusEvents:        tty && term != "dumb" && !screen,
		Tmux:               tmux,
		Screen:             screen,
		SSH:                ssh,
		Windows:            windows,
		Program:            program,
	}
}

File: ports/rust/src/capabilities.rs
//! What the terminal can actually do. Detection is deliberately conservative:
//! we degrade colors and glyphs rather than print mojibake on someone's console.

use std::collections::HashMap;
use std::io::IsTerminal;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub enum ColorDepth {
    #[default]
    TrueColor,
    Ansi256,
    Ansi16,
    None,
}

impl ColorDepth {
    pub fn as_str(self) -> &'static str {
        match self {
            ColorDepth::TrueColor => "truecolor",
            ColorDepth::Ansi256 => "ansi256",
            ColorDepth::Ansi16 => "ansi16",
            ColorDepth::None => "none",
        }
    }

    pub fn parse(s: &str) -> Option<ColorDepth> {
        match s {
            "truecolor" => Some(ColorDepth::TrueColor),
            "ansi256" => Some(ColorDepth::Ansi256),
            "ansi16" => Some(ColorDepth::Ansi16),
            "none" => Some(ColorDepth::None),
            _ => None,
        }
    }
}

#[derive(Clone, Debug)]
pub struct Capabilities {
    /// stdout is a real TTY, not a pipe or a file.
    pub tty: bool,
    pub colors: ColorDepth,
    pub true_color: bool,
    pub unicode: bool,
    pub braille: bool,
    pub mouse: bool,
    /// DEC 2026 atomic frame updates.
    pub synchronized_output: bool,
    pub bracketed_paste: bool,
    pub focus_events: bool,
    pub tmux: bool,
    pub screen: bool,
    pub ssh: bool,
    pub windows: bool,
    /// Best guess at the emulator: kitty, wezterm, ghostty, iterm, alacritty,
    /// vscode, windows-terminal, xterm, unknown.
    pub program: String,
}

impl Default for Capabilities {
    fn default() -> Capabilities {
        Capabilities {
            tty: true,
            colors: ColorDepth::TrueColor,
            true_color: true,
            unicode: true,
            braille: true,
            mouse: true,
            synchronized_output: false,
            bracketed_paste: true,
            focus_events: true,
            tmux: false,
            screen: false,
            ssh: false,
            windows: cfg!(windows),
            program: "unknown".to_string(),
        }
    }
}

/// Force a capability rather than detecting it. Used by tests and by apps that
/// know better than the environment does.
#[derive(Clone, Debug, Default)]
pub struct CapabilityOverrides {
    pub colors: Option<ColorDepth>,
    pub unicode: Option<bool>,
    pub braille: Option<bool>,
    pub mouse: Option<bool>,
    pub synchronized_output: Option<bool>,
    pub tty: Option<bool>,
}

/// The environment, as a map, so detection is testable without touching the
/// real process environment.
pub type Env = HashMap<String, String>;

pub fn process_env() -> Env {
    std::env::vars().collect()
}

fn get<'a>(env: &'a Env, key: &str) -> Option<&'a str> {
    env.get(key).map(String::as_str)
}

fn has(env: &Env, key: &str) -> bool {
    env.get(key).is_some_and(|v| !v.is_empty())
}

pub fn detect_program(env: &Env) -> String {
    let term = get(env, "TERM").unwrap_or("");
    let program = get(env, "TERM_PROGRAM").unwrap_or("");
    if has(env, "KITTY_WINDOW_ID") || term == "xterm-kitty" {
        return "kitty".into();
    }
    if has(env, "WEZTERM_EXECUTABLE") || program == "WezTerm" {
        return "wezterm".into();
    }
    if has(env, "GHOSTTY_RESOURCES_DIR") || term == "xterm-ghostty" {
        return "ghostty".into();
    }
    if program == "iTerm.app" {
        return "iterm".into();
    }
    if has(env, "ALACRITTY_WINDOW_ID") || term == "alacritty" {
        return "alacritty".into();
    }
    if program == "vscode" {
        return "vscode".into();
    }
    if has(env, "WT_SESSION") {
        return "windows-terminal".into();
    }
    if program == "Apple_Terminal" {
        return "apple-terminal".into();
    }
    if has(env, "KONSOLE_VERSION") {
        return "konsole".into();
    }
    if term.starts_with("xterm") {
        return "xterm".into();
    }
    "unknown".into()
}

const TRUECOLOR_PROGRAMS: &[&str] =
    &["kitty", "wezterm", "ghostty", "iterm", "vscode", "windows-terminal", "konsole"];

fn detect_colors(env: &Env, tty: bool) -> ColorDepth {
    if env.get("NO_COLOR").is_some_and(|v| !v.is_empty()) {
        return ColorDepth::None;
    }
    match get(env, "FORCE_COLOR") {
        Some("0") | Some("false") => return ColorDepth::None,
        Some("1") => return ColorDepth::Ansi16,
        Some("2") => return ColorDepth::Ansi256,
        Some("3") => return ColorDepth::TrueColor,
        _ => {}
    }
    // Any other non-empty value — `FORCE_COLOR=true` is the common one —
    // asserts that color works. It is a floor, not a ceiling: returning a level
    // here would cap a truecolor terminal at 16 colors. It only waives the tty
    // check.
    let forced = env.get("FORCE_COLOR").is_some_and(|v| !v.is_empty());
    if !tty && !forced {
        return ColorDepth::None;
    }
    let term = get(env, "TERM").unwrap_or("");
    if term == "dumb" {
        return ColorDepth::None;
    }
    let colorterm = get(env, "COLORTERM").unwrap_or("").to_ascii_lowercase();
    if colorterm == "truecolor" || colorterm == "24bit" {
        return ColorDepth::TrueColor;
    }
    if TRUECOLOR_PROGRAMS.contains(&detect_program(env).as_str()) {
        return ColorDepth::TrueColor;
    }
    if term.contains("256") {
        return ColorDepth::Ansi256;
    }
    ColorDepth::Ansi16
}

fn detect_unicode(env: &Env) -> bool {
    // A dumb terminal has no glyph repertoire to speak of. The Linux console is
    // not in that category — its default font draws box and block elements
    // perfectly well — so only Braille is withheld from it, below.
    if get(env, "TERM").unwrap_or("") == "dumb" {
        return false;
    }
    let locale = get(env, "LC_ALL")
        .filter(|v| !v.is_empty())
        .or_else(|| get(env, "LC_CTYPE").filter(|v| !v.is_empty()))
        .or_else(|| get(env, "LANG").filter(|v| !v.is_empty()))
        .unwrap_or("");
    let upper = locale.to_ascii_uppercase();
    if upper.contains("UTF8") || upper.contains("UTF-8") {
        return true;
    }
    // Windows Terminal and modern emulators are UTF-8 regardless of locale vars.
    if has(env, "WT_SESSION") || has(env, "TERM_PROGRAM") || has(env, "KITTY_WINDOW_ID") {
        return true;
    }
    if cfg!(windows) {
        has(env, "WT_SESSION")
    } else {
        locale.is_empty()
    }
}

const SYNC_PROGRAMS: &[&str] =
    &["kitty", "wezterm", "ghostty", "iterm", "windows-terminal", "konsole", "alacritty"];

pub fn detect_capabilities(overrides: &CapabilityOverrides) -> Capabilities {
    detect_capabilities_in(overrides, &process_env(), std::io::stdout().is_terminal())
}

pub fn detect_capabilities_in(
    overrides: &CapabilityOverrides,
    env: &Env,
    is_tty: bool,
) -> Capabilities {
    let tty = overrides.tty.unwrap_or(is_tty);
    let term = get(env, "TERM").unwrap_or("").to_string();
    let program = detect_program(env);
    let tmux = has(env, "TMUX") || term.starts_with("tmux") || term.starts_with("screen");
    let screen = term.starts_with("screen") && !has(env, "TMUX");
    let ssh = has(env, "SSH_CLIENT") || has(env, "SSH_TTY") || has(env, "SSH_CONNECTION");
    let colors = overrides.colors.unwrap_or_else(|| detect_colors(env, tty));
    let unicode = overrides.unicode.unwrap_or_else(|| detect_unicode(env));
    let sync_capable = SYNC_PROGRAMS.contains(&program.as_str()) || tmux;

    Capabilities {
        tty,
        colors,
        true_color: colors == ColorDepth::TrueColor,
        unicode,
        // The Linux console draws box and block elements but has no Braille in
        // its default font, which is the one glyph class it genuinely lacks.
        braille: overrides
            .braille
            .unwrap_or(unicode && program != "apple-terminal" && term != "linux"),
        mouse: overrides.mouse.unwrap_or(tty && term != "dumb" && term != "linux"),
        synchronized_output: overrides.synchronized_output.unwrap_or(tty && sync_capable),
        bracketed_paste: tty && term != "dumb",
        focus_events: tty && term != "dumb" && !screen,
        tmux,
        screen,
        ssh,
        windows: cfg!(windows),
        program,
    }
}
